File: xor/xor.cpp
/**
 * T1: Implementati un sistem criptografic bazat pe functia XOR.
 * https://github.com/ucv-cs/Securitatea-sistemelor-informatice
 */

#include <iostream>	 // cout, cin, endl
#include <string>	 // getline()

using namespace std;

bool hex_output = false;

// string int_as_hex(string text) {

// }

// string hex_as_int(string text) {
// }

/**
 * Cripteaza/decripteaza un text cu o cheie, folosind operatia xor.
 *
 * @param text
 * @param key
 * @return textul criptat/decriptat
 */
string xor_string(string text, string key) {
	for (int i = 0; i < text.length(); i++) {
		// pentru fiecare caracter din input realizeaza xor cu fiecare caracter
		// din cheie, repetand, daca e cazul tot continutul cheii
		text[i] = text[i] ^ key[i % key.length()];
	}

	return text;
}

/**
 * Pregateste cin pentru un nou input.
 */
void clear_cin() {
	cin.clear();			 // reseteaza starea bufferului cin
	cin.ignore(1000, '\n');	 // "goleste" bufferul cin
}

/**
 * Executia programului.
 */
int main(int argc, const char* argv[]) {
	// citeste argumentele programului
	for (int i = 1; i < argc; i++) {
		if (string(argv[i]) == "-h") {
			// output reprezentat hexazecimal
			// implicit, reprezentarea este in ASCII
			hex_output = true;
		}
	}

	string plain_text;
	string cipher_text;
	string key;
	string separator = "\n----------------------\n";

	bool flag = true;
	while (flag) {
		// obtine de la utilizator textul in clar si cheia
		cout << "[*] Scrie textul de criptat: ";
		getline(cin, plain_text);
		cout << "[*] Scrie cheia: ";
		while (true) {
			getline(cin, key);
			if (key == "") {
				cout << "[!] Cheia nu poate fi goala: ";
				continue;
			}
			break;
		}

		// afiseaza textul criptat si pe cel decriptat (ca verificare)
		cipher_text = xor_string(plain_text, key);
		cout << "\n[*] Text criptat:   " << cipher_text << endl;
		cout << "[*] Text decriptat: " << xor_string(cipher_text, key) << endl;

		// controlul repetarii
		cout << "\n[?] Continuam? (d / n): ";
		char answer;
		cin >> answer;
		clear_cin();
		flag = (answer == 'd') ? true : false;
		cout << separator;
	}
}

File: xor/xor.py
"""
T1: Implementati un sistem criptografic bazat pe functia XOR.
https://github.com/ucv-cs/Securitatea-sistemelor-informatice
"""
import sys

# setarea implicită de afișare
hex_output = True


def encrypt(text, key):
	"""
	Criptează un text cu o cheie, folosind operația xor.
	@param text
	@param key
	@returns textul criptat
	"""
	result = ""
	if hex_output:
		for i in range(len(text)):
			result += hex(ord(text[i]) ^ ord(key[i % len(key)]))[2:].zfill(2)
	else:
		for i in range(len(text)):
			result += chr(ord(text[i]) ^ ord(key[i % len(key)]))

	return result


def decrypt(text, key):
	"""
	Decriptează un text cu o cheie, folosind operația xor.
	@param text
	@param key
	@returns textul decriptat
	"""
	result = ""
	temp = ""
	if hex_output:
		# dacă outputul este în format hexazecimal, trebuie citite grupuri
		# de câte 2 caractere, apoi transformate în int
		for i in range(len(text)):
			if (i + 1) % 2 == 0:
				s = "0x" + text[i - 1] + text[i]
				temp += chr(int(eval(s)))
		text = temp

	for i in range(len(text)):
		result += chr(ord(text[i]) ^ ord(key[i % len(key)]))

	return result


def main():
	"""
	Execuția programului.
	"""
	# switch pentru afișarea textului criptat în format ASCII sau
	# hexazecimal (implicit)
	# utilizare: py -i affine.py -a
	if "-a" in str(sys.argv):
		global hex_output
		hex_output = False

	separator = "\n----------------------"

	# input de la utilizator
	flag = True
	while flag:
		plain_text = input("[*] Scrie textul de criptat: ")
		while True:
			key = input("[*] Scrie cheia: ")
			if key == "":
				continue
			else:
				break

		# output
		cipher_text = encrypt(plain_text, key)
		print(f"\n[*] Text criptat:   {cipher_text}")
		print(f"[*] Text decriptat: {decrypt(cipher_text, key)}")

		flag = True if input(
		    "\n[?] Continuăm? (d / n): ").lower() == "d" else False
		print(separator)


if __name__ == "__main__":
	main()